lib/dune_file.ml

```diff
@@ -3,39 +3,77 @@ open Import
 module Lang = struct
   type version = int * int
 
-  let compare_version (major, minor) (major', minor') =
-    match Int.compare major major' with
-    | 0 -> Int.compare minor minor'
-    | _ as ord -> ord
-
   let pp_version fmt (major, minor) = Format.fprintf fmt "%d.%d" major minor
 
-  let parse_version s =
-    let err () =
-      Error (`Msg (Printf.sprintf "Invalid dune lang version: %s" s))
-    in
-    match String.split_on_char ~sep:'.' s with
-    | [ major; minor ] -> (
-        match (int_of_string_opt major, int_of_string_opt minor) with
-        | Some major, Some minor -> Ok (major, minor)
-        | _ -> err ())
-    | _ -> err ()
-
-  let parse_stanza s =
-    let content =
-      let open Option.O in
-      String.drop_prefix ~prefix:"(" s >>= String.drop_suffix ~suffix:")"
-      >>| fun content -> String.split_on_char ~sep:' ' content
-    in
-    match content with
-    | Some [ "lang"; "dune"; version ] -> parse_version version
-    | _ -> Error (`Msg (Printf.sprintf "Invalid lang stanza: %s" s))
+  let version_to_string version = Format.asprintf "%a" pp_version version
 
-  let is_stanza s = String.is_prefix ~prefix:"(lang " s
+  let stanza_regexp =
+    let open Re in
+    seq
+      [
+        bol;
+        char '(';
+        rep blank;
+        str "lang";
+        rep1 blank;
+        str "dune";
+        rep1 blank;
+        group (rep1 digit);
+        char '.';
+        group (rep1 digit);
+        rep blank;
+        char ')';
+        rep blank;
+        opt (char '\r');
+        eol;
+      ]
 
-  let duniverse_minimum_version = (1, 11)
+  let from_match group =
+    let major_str = Re.Group.get group 1 in
+    let minor_str = Re.Group.get group 2 in
+    match (int_of_string_opt major_str, int_of_string_opt minor_str) with
+    | Some major, Some minor -> Ok (major, minor)
+    | _ ->
+        Printf.ksprintf
+          (fun msg -> Error (`Msg msg))
+          "Invalid dune lang version: %s.%s" major_str minor_str
+
+  let from_content content =
+    let open Result.O in
+    let cregexp = Re.compile stanza_regexp in
```

Comment:
`compile` is the "expensive" operation, so it makes more sense to do it in the toplevel definition above rather than each time.

Reply:
I thought that once compiled, it has some kind of state and that you can't really re-use it. What is compile doing then?

Reply:
No, that's the other way around: `compile` is turning an AST of regexp into an automaton using some graph science, maybe doing some optimizations on the way. You can find an example in the ocaml-re benchmarks where the regexp is compiled once and executed on many inputs:

https://github.com/ocaml/ocaml-re/blob/42c84b0f39117ece63334995258487db0878653d/benchmarks/benchmark.ml#L107-L111